### ISSUE TITLE
Settings: Webhook Setup

### ISSUE THREAD

**Issue by benhaynes** [directus/app#1058]
# Feature Request

## What problem does this feature solve?

It would be nice to allow setting up webhooks from within the App. Moving the API's config/code setup into the App will help make things easier, adds validation, and gives Cloud users a way to access new features.

## How do you think this should be implemented?

@benhaynes will provide a design and add it here.

## Would you be willing to work on this?

I can work on the design – anyone interested in taking on the code?

**Comment by benhaynes** [comment]
These are the Docs for creating Webhooks... these would need to be dynamically created through the interface: https://docs.directus.io/extensions/hooks.html#create-a-hook

Below is an initial design. Webhooks would be accessible from within Settings. We would use the Item Browse and Item Detail templates for these pages.

![directus 7 webhook setup](https://user-images.githubusercontent.com/522079/47855506-31dd2700-ddbb-11e8-8158-59428f1bec51.png)

**Comment by rijkvanzanten** [comment]
To achieve better clarity/visibility, we are now tracking feature requests within the [Feature Request](https://github.com/directus/app/projects/45) project board.

**This issue being closed does not mean it's not being considered.**

**Comment by philleepflorence** [comment]
This looks nice! Happy to dig in on the API side.

**Comment by philleepflorence** [comment]
Also if we could come up with a JSON payload format to dispatch with every webhook.
Also schema for the collection in the DB.

**Comment by benhaynes** [comment]
Perfect. I will work on updating the Admin/Settings GUI, and here is a first stab at a schema and payload (let's tweak this as needed):

Ref: https://developers.facebook.com/docs/graph-api/webhooks/
Ref: https://github.com/lukethacoder/wp-webhook-deploy-netlify
Ref: https://getcockpit.com/documentation/api/webhooks

### Payload
```
{
  "event": "item.update",
  "payload": [
    {
      "time": 1520383571,
      "data": [
        {
          "id": "123",
          "title": "Hello World",
        }
      ]
    }
  ]
}
```

### Schema
```
directus_webhooks
- id
- status (draft, inactive, active)
- name (we don't need this, just an idea)
- url
- type (JSON default)
- events (CSV of Directus event keys)
- collections (CSV of collection keys or NULL if not used)
- secret (to ensure POST requests sent to the payload URL are from Directus)
```

Event Keys: https://docs.directus.io/extensions/hooks.html#action-hooks

**Comment by philleepflorence** [comment]
@benhaynes would you be able to update the DB with `directus_webhooks` included? Then I can pull?

**Comment by benhaynes** [comment]
Is there a branch for this yet? This will be more time consuming to actually do since we need to create full schema/seed migration files for the change. I would say just mock up the table locally based on the comment I made above... or I can create it and Slack the SQL for that table.

@rijkvanzanten are we thinking that we would add interfaces for this into `directus_fields` and use this with a normal form for the collection?

**Comment by philleepflorence** [comment]
I will work on it locally for the time being and access via Sequel Pro until we have the GUI done and/or migration and update done.

then once I have the PR, I will include the SQL Export for those who wish to test.

**Comment by benhaynes** [comment]
Perfect. Once it's ready/tested we'll write the migrations for it. :)

**Comment by rijkvanzanten** [comment]
We can use the regular item detail page! (Assuming we have all the interfaces we need for this page)

**Comment by benhaynes** [comment]
Awesome! Will we need a way to turn on editing for director's system collections? Or will that just work if you have permission?

**Comment by rijkvanzanten** [comment]
Right now that access is blocked by default. I think we should add a card to settings and have a dedicated route for it eg `/settings/webhooks`. That route can be an alias for `/items/directus_webhooks` though

**Comment by philleepflorence** [comment]
> Right now that access is blocked by default. I think we should add a card to settings and have a dedicated route for it eg /settings/webhooks. That route can be an alias for /items/directus_webhooks though

I approve this message!

**Comment by philleepflorence** [comment]
@benhaynes this is currently only in the App Project Board, but I will build the API portion first, do we need to duplicate to API Board or no?

**Comment by rijkvanzanten** [comment]
Lets do it for consistency and accountabilities sake

**Comment by philleepflorence** [comment]
How do you mean? @rijkvanzanten ...

**Comment by philleepflorence** [comment]
In the meantime, I went ahead and put the v6.x Webhooks on GitHub so we can offer better alternatives or comments on the implementation.

All the files involved are in the repo - I will be porting the code from that to match v7.

REPO: https://github.com/philleepflorence/directus-webhooks

**Comment by rijkvanzanten** [comment]
> do we need to duplicate to API Board or no?

Yes, lets add a new issue to the API board for consistency and accountabilities sake 🙂
